chain/chain/src/garbage_collection.rs

```diff
@@ -1134,133 +1128,104 @@ impl<'a> ChainStoreUpdate<'a> {
     }
 }
 
-/// Returns shards that we tracked in an epoch, given a hash of the last block in the epoch.
-/// The block has to be available, so this function has to be called before gc is run for the block.
+/// Cleans up the state of the parent shard once we stop tracking the epoch where the resharding happened.
 ///
-/// Note that validator ID or shard tracking config could have change since the epoch passed,
-/// so we have to rely on what is stored in the database to figure out tracked shards.
-/// We rely on `TrieChanges` column to preserve what shards this node tracked at that time.
-fn get_potential_shards_for_cleanup(
-    chain_store_update: &ChainStoreUpdate,
-    epoch_manager: &Arc<dyn EpochManagerAdapter>,
-    shard_tracker: &ShardTracker,
+/// When we are GC'ing the last block of the epoch where a resharding happened, we need to cleanup the
+/// state of the parent shard. This is done by iterating through the new shard layout and deleting the
+/// state of the parent shard for each child shard.
+fn gc_parent_shard_after_resharding(
+    chain_store_update: &mut ChainStoreUpdate,
+    epoch_manager: &dyn EpochManagerAdapter,
     block_hash: &CryptoHash,
-) -> Result<Option<Vec<ShardUId>>, Error> {
-    if shard_tracker.tracks_all_shards()
-        || !epoch_manager.is_last_block_in_finished_epoch(block_hash)?
-    {
-        return Ok(None);
+) -> Result<(), Error> {
+    // If we are GC'ing the resharding block, i.e. the last block of the epoch, clear out state for the parent shard
+    if !epoch_manager.will_shard_layout_change(block_hash)? {
+        return Ok(());
     }
-    let block = chain_store_update
-        .get_block(block_hash)
-        .expect("block data is not expected to be already cleaned");
-    let epoch_id = block.header().epoch_id();
-    let shard_layout = epoch_manager.get_shard_layout(epoch_id).expect("epoch id must exist");
-    let mut tracked_shards = vec![];
+
+    let _span = tracing::debug_span!(target: "garbage_collection", "gc_resharding").entered();
+
+    // Given block_hash is the resharding block, shard_layout is the shard layout of the next epoch
+    let shard_layout = epoch_manager.get_shard_layout_from_prev_block(block_hash)?;
+    let store = chain_store_update.store();
+    let mut trie_store_update = store.trie_store().store_update();
+    for parent_shard_uid in shard_layout.get_split_parent_shard_uids()? {
+        // Delete the state of the parent shard
+        tracing::info!(target: "garbage_collection", ?parent_shard_uid, "resharding state cleanup");
+        trie_store_update.delete_shard_uid_prefixed_state(parent_shard_uid);
```

Comment:
sanity check - this should not be called on the archival nodes

Reply:
Yes, from my understanding, GC code only runs on normal nodes.

Reply:
Doesn't GC run *differently* on archival nodes? 
It's like a normal node GC with the addition that there's a copy loop to cold DB. Plus GC is stalling to proceed if cold DB lag is high.

I might be confusing the high level vision with the impl details though..

Reply:
Hmm, I might lack the understanding of cold storage loop. I think it's better to take this discussion offline.